### PR TITLE
[Feat] implement RandomRotation3D and RandomAffine3D

### PR DESCRIPTION
**Pull request by shijianjian**
### Description

For a record, I will put the shear matrix calculation here:
Screenshot 2020-08-07 at 12 48 44 AM

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/geometry/transform/imgwarp.py [hunk 1]
```diff
@@ -158,6 +158,68 @@ def warp_affine(src: torch.Tensor, M: torch.Tensor,
                          padding_mode=padding_mode)
 
 
+def warp_affine3d(src: torch.Tensor, M: torch.Tensor,
```
Comment (edgarriba) [inline kornia/geometry/transform/imgwarp.py:161]: @shijianjian this is already warp_projective, right ?

Comment (shijianjian) [reply kornia/geometry/transform/imgwarp.py:161]: Right. I will clean it up. I did not use those functions here.

Comment (shijianjian) [reply kornia/geometry/transform/imgwarp.py:161]: BTW, ```warp_perspective``` is not used for the 2D ```affine()```. Is there any specific reason for that?

File: kornia/geometry/conversions.py [hunk 2]
```diff
@@ -135,6 +136,25 @@ def convert_affinematrix_to_homography(A: torch.Tensor) -> torch.Tensor:
     return H
 
 
+def convert_affinematrix_to_homography3d(A: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/conversions.py:139]: I think this one can be reused from the previous one, only needs a better naming. Suggestions ? @ducha-aiki @lferraz

File: kornia/geometry/transform/affwarp.py [hunk 3]
```diff
@@ -35,6 +39,18 @@ def _compute_tensor_center(tensor: torch.Tensor) -> torch.Tensor:
     return center
 
 
+def _compute_tensor_center3d(tensor: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/transform/affwarp.py:42]: Add small assert `assert len(tensor)>=3`

File: kornia/geometry/transform/affwarp.py [hunk 4]
```diff
@@ -113,6 +150,43 @@ def affine(tensor: torch.Tensor, matrix: torch.Tensor, mode: str = 'bilinear',
     return warped
 
 
+def affine3d(tensor: torch.Tensor, matrix: torch.Tensor, mode: str = 'bilinear',
+             align_corners: bool = False) -> torch.Tensor:
+    r"""Apply an affine transformation to the 3d volume.
+
+    Args:
+        tensor (torch.Tensor): The image tensor to be warped.
```
Comment (edgarriba) [inline kornia/geometry/transform/affwarp.py:158]: Specify shapes

File: kornia/geometry/transform/affwarp.py [hunk 5]
```diff
@@ -150,6 +224,42 @@ def rotate(tensor: torch.Tensor, angle: torch.Tensor,
     return affine(tensor, rotation_matrix[..., :2, :3], mode, align_corners)
 
 
+def rotate3d(tensor: torch.Tensor, yaw: torch.Tensor, pitch: torch.Tensor, roll: torch.Tensor,
```
Comment (edgarriba) [inline kornia/geometry/transform/affwarp.py:227]: Wouldn't be clean to pass *,3 tensor with angles? @ducha-aiki

File: kornia/geometry/transform/projwarp.py [hunk 6]
```diff
@@ -43,10 +43,10 @@ def warp_projective(src: torch.Tensor,
     size_src: Tuple[int, int, int] = (D, H, W)
     size_out: Tuple[int, int, int] = dsize
 
-    M_4x4 = matrix_to_homogeneous(M)  # Bx4x4
+    M_4x4 = convert_affinematrix_to_homography3d(M)  # Bx4x4
```
Comment (edgarriba) [inline kornia/geometry/transform/projwarp.py:46]: I think upper name is more explicative . @ducha-aiki

File: kornia/geometry/warp/homography_warper.py [hunk 7]
```diff
@@ -264,3 +292,41 @@ def normalize_homography(dst_pix_trans_src_pix: torch.Tensor,
         dst_norm_trans_dst_pix @ (dst_pix_trans_src_pix @ src_pix_trans_src_norm)
     )
     return dst_norm_trans_src_norm
+
+
+def normalize_homography3d(dst_pix_trans_src_pix: torch.Tensor,
```
Comment (edgarriba) [inline kornia/geometry/warp/homography_warper.py:297]: We should move this to another place. This are not 3x3  homographies anymore. Maybe to kornia.geometry.linalg ? @ducha-aiki

File: kornia/geometry/transform/affwarp.py [hunk 8]
```diff
@@ -113,6 +150,43 @@ def affine(tensor: torch.Tensor, matrix: torch.Tensor, mode: str = 'bilinear',
...
+    # we enforce broadcasting since by default grid_sample it does not
+    # give support for that
+    matrix = matrix.expand(tensor.shape[0], -1, -1)
+
+    # warp the input tensor
+    depth: int = tensor.shape[-3]
+    height: int = tensor.shape[-2]
+    width: int = tensor.shape[-1]
+    warped: torch.Tensor = warp_projective(tensor, matrix, (depth, height, width), mode,
+                                           align_corners=align_corners)
```
Comment (edgarriba) [inline kornia/geometry/transform/affwarp.py:181]: I open a discussion here about naming `warp_projective` Vs `warp_affine3d` @ducha-aiki @lferraz 

Probably given the current functionality it's the second one but we could include also projection matrix adaption, which in fact I guess would be something similar to `warp_frame_depth`.

File: kornia/geometry/conversions.py (hunk 2 above)
Comment (shijianjian) [reply kornia/geometry/conversions.py:139]: I will put the implementation into ```_convert_affinematrix_to_homography_impl```. Those functions will be left with shape validation only.

**Comment by shijianjian** [comment]
One thing left for the ```RandomAffine3D``` is the translation for the shearing (where placed with zeros):
```python
    m00, m01, m02 = ones, -sxy_tan, -sxz_tan
    m10, m11, m12 = -syx_tan, sxy_tan * syx_tan + ones, sxz_tan * syx_tan - syz_tan
    m20 = syx_tan * szy_tan - szx_tan
    m21 = sxy_tan * szx_tan - szy_tan * m11
    m22 = sxz_tan * szx_tan - szy_tan * m12 +ones
    # TODO: translation math
    shear_mat = torch.stack([
        m00, m10, m20, zeros,
        m01, m11, m21, zeros,
        m02, m12, m22, zeros
    ], dim=-1).view(-1, 3, 4)
```
Other parts should be finished.

File: kornia/augmentation/augmentation3d.py [hunk 9]
```diff
@@ -186,3 +187,144 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+    Args:
+        degrees (float or tuple or list): Range of degrees to select from.
+            If degrees is a number, then yaw, pitch, roll will be generated from the range of (-degrees, +degrees).
+            If degrees is a tuple of (min, max), then yaw, pitch, roll will be generated from the range of (min, max).
+            If degrees is a list of floats [a, b, c], then yaw, pitch, roll will be generated from (-a, a), (-b, b)
+            and (-c, c).
+            If degrees is a list of tuple ((a, b), (m, n), (x, y)), then yaw, pitch, roll will be generated from
+            (a, b), (m, n) and (x, y).
+            Set to 0 to deactivate rotations.
+        translate (tuple, optional): tuple of maximum absolute fraction for horizontal
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentation3d.py:204]: Why translation, scale and share only in 2 directions instead of 3?

File: kornia/augmentation/augmentation3d.py [hunk 10]
```diff
@@ -186,3 +187,144 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+        self.degrees = degrees
+        self.translate = translate
+        self.scale = scale
+        self.shear = shear
+        self.resample: Resample = Resample.get(resample)
+        self.same_on_batch = same_on_batch
+        self.align_corners = align_corners
+
+    def __repr__(self) -> str:
+        repr = f"(degrees={self.degrees}, translate={self.translate}, scale={self.scale}, shear={self.shear}, "
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentation3d.py:245]: add align_corners to repr

File: kornia/augmentation/augmentation3d.py [hunk 11]
```diff
@@ -186,3 +187,144 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
+    ) -> None:
+        super(RandomRotation3D, self).__init__(return_transform)
+        self.degrees = degrees
+        self.interpolation: Resample = Resample.get(interpolation)
+        self.same_on_batch = same_on_batch
+        self.align_corners = align_corners
+
+    def __repr__(self) -> str:
+        repr = f"(degrees={self.degrees}, interpolation={self.interpolation.name}, "
+        f"return_transform={self.return_transform}, same_on_batch={self.same_on_batch})"
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentation3d.py:319]: add align_corners

File: kornia/augmentation/functional3d.py [hunk 12]
```diff
@@ -222,3 +232,135 @@ def compute_dflip_transformation3d(input: torch.Tensor, params: Dict[str, torch.
...
+    Args:
+        input (torch.Tensor): Tensor to be transformed with shape (D, H, W), (C, D, H, W), (B, C, D, H, W).
+        params (Dict[str, torch.Tensor]):
+            - params['angles']: Degrees of rotation with the shape of :math: `(*, 3)` for yaw, pitch, roll.
+            - params['translations']: Horizontal and vertical translations.
+            - params['center']: Rotation center.
+            - params['scale']: Scaling params.
+            - params['sx']: Shear param toward x-axis.
+            - params['sy']: Shear param toward y-axis.
+            - params['sz']: Shear param toward y-axis.
```
Comment (ducha-aiki) [inline kornia/augmentation/functional3d.py:249]: typo: - params['sz']: Shear param toward y-axis. -> - params['sz']: Shear param toward z-axis.

File: kornia/augmentation/functional3d.py [hunk 13]
```diff
@@ -222,3 +232,135 @@ def compute_dflip_transformation3d(input: torch.Tensor, params: Dict[str, torch.
...
+
+
+def apply_affine3d(input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torch.Tensor:
+    r"""Random affine transformation of the image keeping center invariant.
+
+    Args:
+        input (torch.Tensor): Tensor to be transformed with shape (D, H, W), (C, D, H, W), (B, C, D, H, W).
+        params (Dict[str, torch.Tensor]):
+            - params['angles']: Degrees of rotation with the shape of :math: `(*, 3)` for yaw, pitch, roll.
+            - params['translations']: Horizontal and vertical translations.
```
Comment (ducha-aiki) [inline kornia/augmentation/functional3d.py:244]: add z direction

File: kornia/augmentation/functional3d.py [hunk 14]
```diff
@@ -222,3 +232,135 @@ def compute_dflip_transformation3d(input: torch.Tensor, params: Dict[str, torch.
...
+            - params['center']: Rotation center.
+            - params['scale']: Scaling params.
```
Comment (ducha-aiki) [inline kornia/augmentation/functional3d.py:246]: add z direction

File: kornia/augmentation/augmentation3d.py [hunk 15]
```diff
@@ -186,3 +187,144 @@ def compute_transformation(self, input: torch.Tensor, params: Dict[str, torch.Te
...
         return F.apply_dflip3d(input, params)
+
+
+class RandomAffine3D(AugmentationBase3D):
+    r"""Random 3D affine transformation of the image keeping center invariant.
+
+    Args:
+        degrees (float or tuple or list): Range of degrees to select from.
+            If degrees is a number, then yaw, pitch, roll will be generated from the range of (-degrees, +degrees).
+            If degrees is a tuple of (min, max), then yaw, pitch, roll will be generated from the range of (min, max).
```
Comment (ducha-aiki) [inline kornia/augmentation/augmentation3d.py:198]: Please, specify, yaw, pitch and roll are rotation around which axis respectively

File: kornia/augmentation/random_generator3d.py [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+
+def random_affine_generator3d(
+    batch_size: int,
+    depth: int,
+    height: int,
+    width: int,
+    degrees: Union[torch.Tensor, float, Tuple[float, float], Tuple[float, float, float],
+        Tuple[Tuple[float, float], Tuple[float, float], Tuple[float, float]]],
+    translate: Optional[Tuple[float, float, float]] = None,
+    scale: Optional[Tuple[float, float]] = None,
```
Comment (ducha-aiki) [inline kornia/augmentation/random_generator3d.py:62]: scale and shear -> 3 tuples

File: kornia/augmentation/random_generator3d.py [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+
+
+def _get_random_affine_params(
+    batch_size: int,
+    depth: int,
+    height: int,
+    width: int,
+    degrees: torch.Tensor,
+    translate: Optional[Tuple[float, float, float]],
+    scales: Optional[Tuple[float, float]],
```
Comment (ducha-aiki) [inline kornia/augmentation/random_generator3d.py:142]: 3 tuples

File: kornia/geometry/transform/affwarp.py (hunk 8 above)
Comment (ducha-aiki) [reply kornia/geometry/transform/affwarp.py:186]: I think, it is better to have them separated, so warp affine3d. Because affine_grid has efficient pytorch implementation and not the projectvie grid.

File: kornia/geometry/transform/affwarp.py (hunk 5 above)
Comment (ducha-aiki) [reply kornia/geometry/transform/affwarp.py:232]: I don't have strong opinion on it. Probably, 3 with explicit names is easier to understand

File: kornia/geometry/warp/homography_warper.py (hunk 7 above)
Comment (ducha-aiki) [reply kornia/geometry/warp/homography_warper.py:297]: I think, better to have all homographies in a single place actually

Comment (edgarriba) [reply kornia/geometry/warp/homography_warper.py:297]: now have also `kornia.geometry.homography` as well :)

File: kornia/geometry/transform/affwarp.py (hunk 8 above)
Comment (edgarriba) [reply kornia/geometry/transform/affwarp.py:186]: great. @shijianjian feel free to rename and so.

File: kornia/augmentation/augmentation3d.py (hunk 9 above)
Comment (shijianjian) [reply kornia/augmentation/augmentation3d.py:204]: translation and shear are documentation errors. translation got 3 direction and shear got actually 6 directions (S_xy, S_xz, S_yx, S_yz, S_zx, Szy). I will update them later.

Scale factor is implemented as isotropic as ```Affine2D``` version.

File: test/augmentation/test_augmentation_3d.py [hunk 18]
```diff
@@ -0,0 +1,748 @@
...
+    def test_random_hflip(self, device):
+
+        f = RandomHorizontalFlip3D(p=1.0, return_transform=True)
+        f1 = RandomHorizontalFlip3D(p=0., return_transform=True)
+        f2 = RandomHorizontalFlip3D(p=1.)
+        f3 = RandomHorizontalFlip3D(p=0.)
+
+        input = torch.tensor([[[0., 0., 0., 0.],
+                               [0., 0., 0., 0.],
+                               [0., 0., 1., 2.]]])  # 1 x 3 x 4
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_3d.py:38]: I would try for D > 1 too

Comment (edgarriba) [reply test/augmentation/test_augmentation_3d.py:38]: besides, what's the input shape expectation fro 3d transforms ? I would assume BxCxDxHxW all the time

File: test/augmentation/test_augmentation_3d.py [hunk 19]
```diff
@@ -0,0 +1,748 @@
...
+
+        expected_transform_1 = expected_transform @ expected_transform
+        expected_transform_1 = expected_transform_1.to(device)
+
+        assert(f(input)[0] == input).all()
+        assert(f(input)[1] == expected_transform_1).all()
+        assert(f1(input)[0] == input).all()
+        assert(f1(input)[1] == expected_transform).all()
+
+    @pytest.mark.skip(reason="turn off all jit for a while")
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_3d.py:141]: we should start enabling jit test in the next release. Will add as item in the roadmap

File: test/augmentation/test_augmentation_3d.py (hunk 18 above)
Comment (edgarriba) [inline test/augmentation/test_augmentation_3d.py:38]: BTW, I think this is wrong. The flip is happening in the vertical axis

Comment (shijianjian) [reply test/augmentation/test_augmentation_3d.py:38]: For 2D, we enabled HW, CHW, BCHW.

For 3D, we enabled DHW, CDHW, BCDHW.

Comment (shijianjian) [reply test/augmentation/test_augmentation_3d.py:38]: What is the definition here? I used:
- Horizontal: flip(-1)
- Vertical: flip(-2)
- Depthical: flip(-3)

Comment (edgarriba) [reply test/augmentation/test_augmentation_3d.py:38]: ok, then is fine I guess

Comment (edgarriba) [reply test/augmentation/test_augmentation_3d.py:38]: ok, make sure this is well documented

Comment (shijianjian) [reply test/augmentation/test_augmentation_3d.py]: > I would try for D > 1 too

I have updated it to D=2 case. It should be alright to merge now.
